### PR TITLE
feat: instant meeting browser notifcations

### LINKED ISSUE #16373: [CAL-4177] Browser Notifications For Instant Meetings
<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-4177](https://linear.app/calcom/issue/CAL-4177/browser-notifications-for-instant-meetings)</sub>

### PR DESCRIPTION
**Pull request by Udit-takkar**
## What does this PR do?

- Fixes https://github.com/calcom/cal.com/issues/16373 (GitHub issue number)
- Fixes CAL-4177(Linear issue number - should be visible at the bottom of the GitHub issue description)

https://github.com/user-attachments/assets/e1959d74-dfc2-437c-ab3b-ad6f0ee944b2

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] N/A I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

Testing it locally is a bit difficult as browser notification only work on https domain. There is a nice article [here](https://vercel.com/guides/access-nextjs-localhost-https-certificate-self-signed) on how we can do that

- Create a new public-private key pair required for web push notifications using `npx web-push generate-vapid-keys`.
- Add these keys in .env file `NEXT_PUBLIC_VAPID_PUBLIC_KEY` and `VAPID_PRIVATE_KEY`.
-  Login in to an org account (owner1-dunder@example.com) 
-  execute ngrok http 3000
-  Now open dev console and copy next auth cookies from app.cal.local to ngrok domain (the idea is to be logged in as owner1-dunder@example.com on ngrok domain)
-  Enable Browser Notification on `bookings/upcoming`

https://github.com/user-attachments/assets/2bef76dd-09a0-4109-93cb-70c4ab1a766b

Now instant meeting event setup

- Create any org team event type 
- Enable Instant bookings toggle
- Go to the booking page and click on connect now bottom modal

You should see a notification which when clicked would redirect you to /connect-and-join page

### CODE REVIEW

File: packages/features/shell/Shell.tsx [hunk 1]
```diff
@@ -1084,7 +1085,9 @@ function SideBar({ bannersHeight, user }: SideBarProps) {
 
 export function ShellMain(props: LayoutProps) {
   const router = useRouter();
-  const { isLocaleReady } = useLocale();
+  const { isLocaleReady, t } = useLocale();
+
+  const { buttonToShow, isLoading, enableNotifications, disableNotifications } = useNotifications();
```
Comment (Udit-takkar) [inline packages/features/shell/Shell.tsx:1090]: Moved it here to enable receiving subscription

File: apps/web/public/service-worker.js [hunk 2]
```diff
@@ -1,16 +1,68 @@
...
+  const allClients = await clients.matchAll({
+    type: 'window',
+    includeUncontrolled: true
+  });
+
+  if(!allClients.length) {
+    console.log("No open tabs, skipping the push notification.");
+    return;
+  }
+
```
Comment (Udit-takkar) [inline apps/web/public/service-worker.js:13]: We only want to display notification when app.cal.com is open (focussed or not)

File: apps/web/public/service-worker.js [hunk 3]
```diff
@@ -1,16 +1,68 @@
...
+    event.waitUntil(self.clients.openWindow(url));
+  }
+
+  switch (event.action) {
+    case 'connect-action':
+      event.notification.close();
+      const url = event.notification.data.url;
+      event.waitUntil(self.clients.openWindow(url));
+      break;
+  }
```
Comment (Udit-takkar) [inline apps/web/public/service-worker.js:67]: To add more actions later based on requirements

File: apps/web/public/service-worker.js [hunk 4]
```diff
@@ -1,16 +1,68 @@
...
+      vibrate: notification.vibrate,
+      requireInteraction: notification.requireInteraction,
+      tag: notification.tag,
+    };
+    self.registration.showNotification(notification.title, options);
+  });
+
+
+  // Show the new notification
+  self.registration.showNotification(title, newNotificationOptions);
```
Comment (Udit-takkar) [inline apps/web/public/service-worker.js:50]: To basically group all notifications to expand or respond to

File: packages/features/instant-meeting/handleInstantMeeting.ts [hunk 5]
```diff
@@ -86,6 +89,69 @@ const handleInstantMeetingWebhookTrigger = async (args: {
...
+          NotificationsSubscriptions: {
+            select: {
+              id: true,
+              subscription: true,
+            },
+          },
+        },
+      },
+    },
+  });
```
Comment (alishaz-polymath) [inline packages/features/instant-meeting/handleInstantMeeting.ts:117]: NIT: Could we not do a trpc loggedInViewer here? Would probably be the better route.

File: packages/features/instant-meeting/handleInstantMeeting.ts [hunk 6]
```diff
@@ -252,6 +318,12 @@ async function handler(req: NextApiRequest) {
     teamId: eventType.team?.id,
   });
 
+  await triggerBrowserNotifications({
+    title: newBooking.title,
+    connectAndJoinUrl: webhookData.connectAndJoinUrl,
+    teamId: eventType.team?.id,
```
Comment (alishaz-polymath) [inline packages/features/instant-meeting/handleInstantMeeting.ts:324]:
```
const triggerBrowserNotifications = async (args: {
  title: string;
  connectAndJoinUrl: string;
  teamId: number;
})
```
this expects teamId to be a number, but teamId here can be undefined

File: packages/features/shell/Shell.tsx [hunk 7]
```diff
@@ -1144,6 +1147,23 @@ export function ShellMain(props: LayoutProps) {
                 </div>
               )}
               {props.actions && props.actions}
+              {props.heading === "Bookings" && buttonToShow && (
+                <Button
+                  color="primary"
+                  onClick={buttonToShow === ButtonState.ALLOW ? enableNotifications : disableNotifications}
+                  loading={isLoading}
+                  disabled={buttonToShow === ButtonState.DENIED}
```
Comment (alishaz-polymath) [inline packages/features/shell/Shell.tsx:1155]: Not part of this PR but I think the ButtonState enums are not consistent
```
export enum ButtonState {
  NONE = "none",
  ALLOW = "allow",
  DISABLE = "disable",
  DENIED = "denied",
}
``` 
in `package/lib/hooks`, we should probably update DENIED to DENY to be consistent. Again, super NIT that is not really part of this PR, but is related.

File: packages/trpc/server/routers/loggedInViewer/addNotificationsSubscription.handler.ts [hunk 8]
```diff
@@ -10,20 +15,52 @@ type AddSecondaryEmailOptions = {
...
 
   if (!existingSubscription) {
     await prisma.notificationsSubscriptions.create({
       data: { userId: user.id, subscription },
     });
+  } else {
+    await prisma.notificationsSubscriptions.update({
+      where: { id: existingSubscription.id },
+      data: { userId: user.id, subscription },
+    });
```
Comment (alishaz-polymath) [inline packages/trpc/server/routers/loggedInViewer/addNotificationsSubscription.handler.ts:46]: Does this mean a user can only have single subscription? 🤔

**Review by alishaz-polymath — COMMENTED** [review]
Left a few NITs for your consideration, would test it next 🙏

File: packages/trpc/server/routers/loggedInViewer/addNotificationsSubscription.handler.ts (hunk 8 above)
Comment (Udit-takkar) [reply packages/trpc/server/routers/loggedInViewer/addNotificationsSubscription.handler.ts:46]: yeah right now just one.

File: packages/features/instant-meeting/handleInstantMeeting.ts (hunk 5 above)
Comment (Udit-takkar) [reply packages/features/instant-meeting/handleInstantMeeting.ts:117]: > Could we not do a trpc loggedInViewer here? 

Could you elaborate more on this?

For more context when someone submits a instant booking meeting handleInstantMeeting is executed (just like handleNewBooking for normal event types) and many other tasks are executed like triggering webhook, creating video call url. etc.

File: apps/web/public/service-worker.js [hunk 9]
```diff
@@ -1,16 +1,68 @@
...
-  const image ="/cal-com-icon.svg";
-  const options = {
-    ...notificationData.options,
+  const allClients = await clients.matchAll({
+    type: 'window',
+    includeUncontrolled: true
+  });
+
+  if(!allClients.length) {
+    console.log("No open tabs, skipping the push notification.");
```
[the remark this answers is not in the document]
Comment (Udit-takkar) [reply apps/web/public/service-worker.js:10]: This is only run on client side so console.log is fine

**Review by alishaz-polymath — DISMISSED** [review]
Tested on Chrome, works 🚀

**Comment by Udit-takkar** [comment]
TODO: Type err.

**Review by alishaz-polymath — APPROVED** [review]

File: packages/features/shell/Shell.tsx [hunk 10]
```diff
@@ -1144,6 +1147,23 @@ export function ShellMain(props: LayoutProps) {
                 </div>
               )}
               {props.actions && props.actions}
+              {props.heading === "Bookings" && buttonToShow && (
```
Comment (hbjORbj) [inline packages/features/shell/Shell.tsx:1150]: What if heading is i18n string?
